Verify service is bound before attempting to disconnect

### Description

Calling `LostApiClientImpl.disconnect()` can result in fatal crash if service is not bound.

### Steps to Reproduce

Attempt to disconnect client before fully connected.

### Lost & Android Version

Lost 2.1.1
Android 6.0.1

```
10-21 17:26:00.092 24813-24813/? E/AndroidRuntime: FATAL EXCEPTION: main
                                                   Process: com.mapzen.erasermap, PID: 24813
                                                   java.lang.RuntimeException: Unable to pause activity {com.mapzen.erasermap/com.mapzen.erasermap.controller.MainActivity}: java.lang.IllegalArgumentException: Service not registered: com.mapzen.android.lost.internal.FusedLocationProviderApiImpl$2@c390cb5
                                                       at android.app.ActivityThread.performPauseActivity(ActivityThread.java:3381)
                                                       at android.app.ActivityThread.performPauseActivity(ActivityThread.java:3340)
                                                       at android.app.ActivityThread.handlePauseActivity(ActivityThread.java:3315)
                                                       at android.app.ActivityThread.access$1100(ActivityThread.java:150)
                                                       at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1355)
                                                       at android.os.Handler.dispatchMessage(Handler.java:102)
                                                       at android.os.Looper.loop(Looper.java:148)
                                                       at android.app.ActivityThread.main(ActivityThread.java:5417)
                                                       at java.lang.reflect.Method.invoke(Native Method)
                                                       at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:726)
                                                       at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:616)
                                                    Caused by: java.lang.IllegalArgumentException: Service not registered: com.mapzen.android.lost.internal.FusedLocationProviderApiImpl$2@c390cb5
                                                       at android.app.LoadedApk.forgetServiceDispatcher(LoadedApk.java:1044)
                                                       at android.app.ContextImpl.unbindService(ContextImpl.java:1337)
                                                       at android.content.ContextWrapper.unbindService(ContextWrapper.java:616)
                                                       at com.mapzen.android.lost.internal.FusedLocationProviderApiImpl$1.onDisconnect(FusedLocationProviderApiImpl.java:59)
                                                       at com.mapzen.android.lost.internal.FusedLocationServiceConnectionManager.disconnect(FusedLocationServiceConnectionManager.java:65)
                                                       at com.mapzen.android.lost.internal.FusedLocationProviderApiImpl.disconnect(FusedLocationProviderApiImpl.java:94)
                                                       at com.mapzen.android.lost.internal.LostApiClientImpl.disconnect(LostApiClientImpl.java:53)
                                                       at com.mapzen.erasermap.model.LostClientManager.disconnect(LostClientManager.kt:55)
                                                       at com.mapzen.erasermap.controller.MainActivity.onPause(MainActivity.kt:217)
                                                       at android.app.Activity.performPause(Activity.java:6403)
                                                       at android.app.Instrumentation.callActivityOnPause(Instrumentation.java:1312)
                                                       at android.app.ActivityThread.performPauseActivity(ActivityThread.java:3367)
                                                       at android.app.ActivityThread.performPauseActivity(ActivityThread.java:3340) 
                                                       at android.app.ActivityThread.handlePauseActivity(ActivityThread.java:3315) 
                                                       at android.app.ActivityThread.access$1100(ActivityThread.java:150) 
                                                       at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1355) 
                                                       at android.os.Handler.dispatchMessage(Handler.java:102) 
                                                       at android.os.Looper.loop(Looper.java:148) 
                                                       at android.app.ActivityThread.main(ActivityThread.java:5417) 
                                                       at java.lang.reflect.Method.invoke(Native Method) 
                                                       at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:726) 
                                                       at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:616) 
```

--- a/lost/src/main/java/com/mapzen/android/lost/internal/ClientManager.java
+++ b/lost/src/main/java/com/mapzen/android/lost/internal/ClientManager.java
@@ -41,7 +41,6 @@ public interface ClientManager {
   void reportProviderDisabled(String provider);
   void notifyLocationAvailability(final LocationAvailability availability);
   boolean hasNoListeners();
-  void disconnect(LostApiClient client);
   void shutdown();
   Map<LostApiClient, Set<LocationListener>> getLocationListeners();
   Map<LostApiClient, Set<PendingIntent>> getPendingIntents();

--- a/lost/src/main/java/com/mapzen/android/lost/internal/FusedLocationProviderApiImpl.java
+++ b/lost/src/main/java/com/mapzen/android/lost/internal/FusedLocationProviderApiImpl.java
@@ -52,10 +52,7 @@ public class FusedLocationProviderApiImpl
       }
     }
 
-    @Override public void onDisconnect(LostApiClient client, boolean disconnectService) {
-      if (disconnectService) {
-        service.disconnect(client);
-      }
+    @Override public void onDisconnect() {
       context.unbindService(serviceConnection);
       Intent intent = new Intent(context, FusedLocationProviderService.class);
       context.stopService(intent);
@@ -90,8 +87,8 @@ public class FusedLocationProviderApiImpl
     serviceConnectionManager.connect(context, callbacks);
   }
 
-  public void disconnect(LostApiClient client) {
-    serviceConnectionManager.disconnect(client);
+  public void disconnect() {
+    serviceConnectionManager.disconnect();
   }
 
   public boolean isConnected() {

--- a/lost/src/main/java/com/mapzen/android/lost/internal/FusedLocationProviderApiImpl.java
+++ b/lost/src/main/java/com/mapzen/android/lost/internal/FusedLocationProviderApiImpl.java
@@ -8,6 +8,7 @@ import com.mapzen.android.lost.api.LocationRequest;
 import com.mapzen.android.lost.api.LostApiClient;
 import com.mapzen.android.lost.api.PendingResult;
 import com.mapzen.android.lost.api.Status;
+import com.mapzen.android.lost.internal.FusedLocationServiceConnectionManager.EventCallbacks;
 
 import android.app.PendingIntent;
 import android.content.ComponentName;
@@ -26,53 +27,52 @@ import java.util.Set;
  * Implementation of the {@link FusedLocationProviderApi}.
  */
 public class FusedLocationProviderApiImpl
-    implements FusedLocationProviderApi {
+    implements FusedLocationProviderApi, EventCallbacks, ServiceConnection {
 
   private Context context;
   private FusedLocationProviderService service;
   private FusedLocationServiceConnectionManager serviceConnectionManager;
+  private boolean isBound;
 
-  private FusedLocationServiceConnectionManager.EventCallbacks eventCallbacks =
-      new FusedLocationServiceConnectionManager.EventCallbacks() {
-    @Override public void onConnect(Context context) {
-      FusedLocationProviderApiImpl.this.context = context;
+  @Override public void onConnect(Context context) {
+    this.context = context;
+    final Intent intent = new Intent(context, FusedLocationProviderService.class);
+    context.startService(intent);
+    context.bindService(intent, this, Context.BIND_AUTO_CREATE);
+  }
 
-      Intent intent = new Intent(context, FusedLocationProviderService.class);
-      context.startService(intent);
+  @Override public void onServiceConnected(IBinder binder) {
+    FusedLocationProviderService.FusedLocationProviderBinder fusedBinder =
+        (FusedLocationProviderService.FusedLocationProviderBinder) binder;
+    if (fusedBinder != null) {
+      service = fusedBinder.getService();
+      isBound = true;
+    }
+  }
 
-      intent = new Intent(context, FusedLocationProviderService.class);
-      context.bindService(intent, serviceConnection, Context.BIND_AUTO_CREATE);
+  @Override public void onDisconnect() {
+    if (isBound) {
+      context.unbindService(this);
     }
 
-    @Override public void onServiceConnected(IBinder binder) {
-      FusedLocationProviderService.FusedLocationProviderBinder fusedBinder =
-          (FusedLocationProviderService.FusedLocationProviderBinder) binder;
-      if (fusedBinder != null) {
-        service = fusedBinder.getService();
-      }
-    }
+    Intent intent = new Intent(context, FusedLocationProviderService.class);
+    context.stopService(intent);
+    service = null;
+  }
 
-    @Override public void onDisconnect() {
-      context.unbindService(serviceConnection);
-      Intent intent = new Intent(context, FusedLocationProviderService.class);
-      context.stopService(intent);
-      service = null;
-    }
-  };
+  @Override public void onServiceConnected(ComponentName name, IBinder binder) {
+    serviceConnectionManager.onServiceConnected(binder);
+    isBound = true;
+  }
 
-  private final ServiceConnection serviceConnection = new ServiceConnection() {
-    @Override public void onServiceConnected(ComponentName name, IBinder binder) {
-      serviceConnectionManager.onServiceConnected(binder);
-    }
-
-    @Override public void onServiceDisconnected(ComponentName name) {
-      serviceConnectionManager.onServiceDisconnected();
-    }
-  };
+  @Override public void onServiceDisconnected(ComponentName name) {
+    serviceConnectionManager.onServiceDisconnected();
+    isBound = false;
+  }
 
   public FusedLocationProviderApiImpl(FusedLocationServiceConnectionManager connectionManager) {
     serviceConnectionManager = connectionManager;
-    serviceConnectionManager.setEventCallbacks(eventCallbacks);
+    serviceConnectionManager.setEventCallbacks(this);
   }
 
   public boolean isConnecting() {

--- a/lost/src/main/java/com/mapzen/android/lost/internal/FusedLocationProviderApiImpl.java
+++ b/lost/src/main/java/com/mapzen/android/lost/internal/FusedLocationProviderApiImpl.java
@@ -53,6 +53,7 @@ public class FusedLocationProviderApiImpl
   @Override public void onDisconnect() {
     if (isBound) {
       context.unbindService(this);
+      isBound = false;
     }
 
     Intent intent = new Intent(context, FusedLocationProviderService.class);

--- a/lost/src/main/java/com/mapzen/android/lost/internal/FusedLocationProviderService.java
+++ b/lost/src/main/java/com/mapzen/android/lost/internal/FusedLocationProviderService.java
@@ -113,8 +113,4 @@ public class FusedLocationProviderService extends Service {
   public Map<LostApiClient, Set<LocationListener>> getLocationListeners() {
     return serviceImpl.getLocationListeners();
   }
-
-  public void disconnect(LostApiClient client) {
-    serviceImpl.disconnect(client);
-  }
 }

--- a/lost/src/main/java/com/mapzen/android/lost/internal/FusedLocationProviderServiceImpl.java
+++ b/lost/src/main/java/com/mapzen/android/lost/internal/FusedLocationProviderServiceImpl.java
@@ -157,10 +157,6 @@ public class FusedLocationProviderServiceImpl implements LocationEngine.Callback
     return clientManager.getLocationCallbacks();
   }
 
-  public void disconnect(LostApiClient client) {
-    clientManager.disconnect(client);
-  }
-
   /**
    * Checks if any listeners or pending intents are still registered for location updates. If not,
    * then shutdown the location engines.

--- a/lost/src/main/java/com/mapzen/android/lost/internal/FusedLocationServiceConnectionManager.java
+++ b/lost/src/main/java/com/mapzen/android/lost/internal/FusedLocationServiceConnectionManager.java
@@ -14,7 +14,7 @@ public class FusedLocationServiceConnectionManager {
   public interface EventCallbacks {
     void onConnect(Context context);
     void onServiceConnected(IBinder binder);
-    void onDisconnect(LostApiClient client, boolean disconnectService);
+    void onDisconnect();
   }
 
   private enum ConnectState { IDLE, CONNECTING, CONNECTED }
@@ -57,12 +57,12 @@ public class FusedLocationServiceConnectionManager {
     addCallbacks(callbacks);
   }
 
-  public void disconnect(LostApiClient client) {
+  public void disconnect() {
     if (connectState != ConnectState.IDLE) {
       boolean disconnectService = (connectState == ConnectState.CONNECTED);
       connectState = ConnectState.IDLE;
       if (eventCallbacks != null) {
-        eventCallbacks.onDisconnect(client, disconnectService);
+        eventCallbacks.onDisconnect();
       }
     }
   }

--- a/lost/src/main/java/com/mapzen/android/lost/internal/FusedLocationServiceConnectionManager.java
+++ b/lost/src/main/java/com/mapzen/android/lost/internal/FusedLocationServiceConnectionManager.java
@@ -59,7 +59,6 @@ public class FusedLocationServiceConnectionManager {
 
   public void disconnect() {
     if (connectState != ConnectState.IDLE) {
-      boolean disconnectService = (connectState == ConnectState.CONNECTED);
       connectState = ConnectState.IDLE;
       if (eventCallbacks != null) {
         eventCallbacks.onDisconnect();

--- a/lost/src/main/java/com/mapzen/android/lost/internal/LostApiClientImpl.java
+++ b/lost/src/main/java/com/mapzen/android/lost/internal/LostApiClientImpl.java
@@ -23,10 +23,12 @@ public class LostApiClientImpl implements LostApiClient {
     if (!geofencingApi.isConnected()) {
       geofencingApi.connect(context);
     }
+
     SettingsApiImpl settingsApi = getSettingsApiImpl();
     if (!settingsApi.isConnected()) {
       settingsApi.connect(context);
     }
+
     FusedLocationProviderApiImpl fusedApi = getFusedLocationProviderApiImpl();
     if (fusedApi.isConnected()) {
       if (connectionCallbacks != null) {
@@ -39,6 +41,7 @@ public class LostApiClientImpl implements LostApiClient {
     } else {
       fusedApi.connect(context, connectionCallbacks);
     }
+
     clientManager.addClient(this);
   }
 
@@ -50,7 +53,7 @@ public class LostApiClientImpl implements LostApiClient {
 
     getSettingsApiImpl().disconnect();
     getGeofencingImpl().disconnect();
-    getFusedLocationProviderApiImpl().disconnect(null);
+    getFusedLocationProviderApiImpl().disconnect();
   }
 
   @Override public boolean isConnected() {

--- a/lost/src/main/java/com/mapzen/android/lost/internal/LostClientManager.java
+++ b/lost/src/main/java/com/mapzen/android/lost/internal/LostClientManager.java
@@ -61,6 +61,10 @@ public class LostClientManager implements ClientManager {
 
   public void removeClient(LostApiClient client) {
     clients.remove(client);
+    clientToListeners.remove(client);
+    clientToPendingIntents.remove(client);
+    clientToLocationCallbacks.remove(client);
+    clientCallbackToLoopers.remove(client);
   }
 
   public int numberOfClients() {
@@ -241,13 +245,6 @@ public class LostClientManager implements ClientManager {
   public boolean hasNoListeners() {
     return clientToListeners.isEmpty() && clientToPendingIntents.isEmpty() &&
         clientToLocationCallbacks.isEmpty();
-  }
-
-  public void disconnect(LostApiClient client) {
-    clientToListeners.remove(client);
-    clientToPendingIntents.remove(client);
-    clientToLocationCallbacks.remove(client);
-    clientCallbackToLoopers.remove(client);
   }
 
   public void shutdown() {

--- a/lost/src/test/java/com/mapzen/android/lost/internal/FusedLocationProviderApiImplTest.java
+++ b/lost/src/test/java/com/mapzen/android/lost/internal/FusedLocationProviderApiImplTest.java
@@ -24,6 +24,7 @@ import java.io.File;
 import static org.fest.assertions.api.Assertions.assertThat;
 import static org.mockito.Matchers.any;
 import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.never;
 import static org.mockito.Mockito.spy;
 import static org.mockito.Mockito.verify;
 import static org.mockito.Mockito.when;
@@ -198,12 +199,6 @@ public class FusedLocationProviderApiImplTest {
   @Test public void onDisconnect_shouldUnbindServiceIfBound() throws Exception {
     Context context = mock(Context.class);
     api.onConnect(context);
-
-    FusedLocationProviderService.FusedLocationProviderBinder binder =
-        mock(FusedLocationProviderService.FusedLocationProviderBinder.class);
-    when(binder.getService()).thenReturn(mock(FusedLocationProviderService.class));
-    api.onServiceConnected(binder);
-
     api.onDisconnect();
     verify(context).unbindService(api);
   }
@@ -211,8 +206,8 @@ public class FusedLocationProviderApiImplTest {
   @Test public void onDisconnect_shouldNotUnbindServiceIfNotBound() throws Exception {
     Context context = mock(Context.class);
     api.onConnect(context);
-
+    api.onServiceDisconnected(mock(ComponentName.class));
     api.onDisconnect();
-    verify(context).unbindService(api);
+    verify(context, never()).unbindService(api);
   }
 }

--- a/lost/src/test/java/com/mapzen/android/lost/internal/FusedLocationProviderApiImplTest.java
+++ b/lost/src/test/java/com/mapzen/android/lost/internal/FusedLocationProviderApiImplTest.java
@@ -26,6 +26,7 @@ import static org.mockito.Matchers.any;
 import static org.mockito.Mockito.mock;
 import static org.mockito.Mockito.never;
 import static org.mockito.Mockito.spy;
+import static org.mockito.Mockito.times;
 import static org.mockito.Mockito.verify;
 import static org.mockito.Mockito.when;
 import static org.robolectric.RuntimeEnvironment.application;
@@ -209,5 +210,13 @@ public class FusedLocationProviderApiImplTest {
     api.onServiceDisconnected(mock(ComponentName.class));
     api.onDisconnect();
     verify(context, never()).unbindService(api);
+  }
+
+  @Test public void onDisconnect_shouldNotAttemptToUnbindServiceMoreThanOnce() throws Exception {
+    Context context = mock(Context.class);
+    api.onConnect(context);
+    api.onDisconnect();
+    api.onDisconnect();
+    verify(context, times(1)).unbindService(api);
   }
 }

--- a/lost/src/test/java/com/mapzen/android/lost/internal/FusedLocationProviderApiImplTest.java
+++ b/lost/src/test/java/com/mapzen/android/lost/internal/FusedLocationProviderApiImplTest.java
@@ -9,12 +9,13 @@ import org.junit.Before;
 import org.junit.Test;
 import org.junit.runner.RunWith;
 import org.mockito.Mockito;
-import org.robolectric.RobolectricGradleTestRunner;
+import org.robolectric.RobolectricTestRunner;
 import org.robolectric.annotation.Config;
 
 import android.app.PendingIntent;
 import android.content.ComponentName;
 import android.content.Context;
+import android.content.Intent;
 import android.location.Location;
 import android.os.Looper;
 
@@ -29,7 +30,7 @@ import static org.mockito.Mockito.when;
 import static org.robolectric.RuntimeEnvironment.application;
 import static org.robolectric.Shadows.shadowOf;
 
-@RunWith(RobolectricGradleTestRunner.class)
+@RunWith(RobolectricTestRunner.class)
 @SuppressWarnings("MissingPermission")
 @Config(constants = BuildConfig.class, sdk = 21, manifest = Config.NONE)
 public class FusedLocationProviderApiImplTest {
@@ -179,5 +180,39 @@ public class FusedLocationProviderApiImplTest {
   @Test public void getListeners() {
     api.getLocationListeners();
     verify(service).getLocationListeners();
+  }
+
+  @Test public void onConnect_shouldStartService() throws Exception {
+    Context context = mock(Context.class);
+    api.onConnect(context);
+    verify(context).startService(new Intent(context, FusedLocationProviderService.class));
+  }
+
+  @Test public void onConnect_shouldBindService() throws Exception {
+    Context context = mock(Context.class);
+    api.onConnect(context);
+    verify(context).bindService(new Intent(context, FusedLocationProviderService.class), api,
+        Context.BIND_AUTO_CREATE);
+  }
+
+  @Test public void onDisconnect_shouldUnbindServiceIfBound() throws Exception {
+    Context context = mock(Context.class);
+    api.onConnect(context);
+
+    FusedLocationProviderService.FusedLocationProviderBinder binder =
+        mock(FusedLocationProviderService.FusedLocationProviderBinder.class);
+    when(binder.getService()).thenReturn(mock(FusedLocationProviderService.class));
+    api.onServiceConnected(binder);
+
+    api.onDisconnect();
+    verify(context).unbindService(api);
+  }
+
+  @Test public void onDisconnect_shouldNotUnbindServiceIfNotBound() throws Exception {
+    Context context = mock(Context.class);
+    api.onConnect(context);
+
+    api.onDisconnect();
+    verify(context).unbindService(api);
   }
 }

--- a/lost/src/test/java/com/mapzen/android/lost/internal/FusedLocationProviderApiImplTest.java
+++ b/lost/src/test/java/com/mapzen/android/lost/internal/FusedLocationProviderApiImplTest.java
@@ -1,11 +1,5 @@
 package com.mapzen.android.lost.internal;
 
-import android.app.PendingIntent;
-import android.content.ComponentName;
-import android.content.Context;
-import android.location.Location;
-import android.os.Looper;
-
 import com.mapzen.android.lost.api.LocationListener;
 import com.mapzen.android.lost.api.LocationRequest;
 import com.mapzen.android.lost.api.LostApiClient;
@@ -17,6 +11,12 @@ import org.junit.runner.RunWith;
 import org.mockito.Mockito;
 import org.robolectric.RobolectricGradleTestRunner;
 import org.robolectric.annotation.Config;
+
+import android.app.PendingIntent;
+import android.content.ComponentName;
+import android.content.Context;
+import android.location.Location;
+import android.os.Looper;
 
 import java.io.File;
 
@@ -94,9 +94,8 @@ public class FusedLocationProviderApiImplTest {
   }
 
   @Test public void disconnect_shouldCallConnectionManager() {
-    LostApiClient client = mock(LostApiClient.class);
-    api.disconnect(client);
-    verify(connectionManager).disconnect(client);
+    api.disconnect();
+    verify(connectionManager).disconnect();
   }
 
   @Test public void isConnected_shouldCallConnectionManager() {

--- a/lost/src/test/java/com/mapzen/android/lost/internal/FusedLocationProviderServiceImplTest.java
+++ b/lost/src/test/java/com/mapzen/android/lost/internal/FusedLocationProviderServiceImplTest.java
@@ -813,36 +813,6 @@ public class FusedLocationProviderServiceImplTest {
     assertThat(otherCallback.getAvailability()).isNull();
   }
 
-  @Test public void disconnect_otherClientShouldHavePendingIntents() {
-    client.connect();
-    otherClient.connect();
-    api.requestLocationUpdates(otherClient, LocationRequest.create(),
-        mock(PendingIntent.class));
-    api.disconnect(client);
-
-    assertThat(api.getPendingIntents().get(otherClient)).isNotNull();
-  }
-
-  @Test public void disconnect_otherClientShouldHaveListeners() {
-    client.connect();
-    otherClient.connect();
-    api.requestLocationUpdates(otherClient, LocationRequest.create(),
-        new TestLocationListener());
-    api.disconnect(client);
-
-    assertThat(api.getLocationListeners().get(otherClient)).isNotNull();
-  }
-
-  @Test public void disconnect_otherClientShouldHaveLocationListeners() {
-    client.connect();
-    otherClient.connect();
-    api.requestLocationUpdates(otherClient, LocationRequest.create(),
-        new TestLocationCallback(), Looper.myLooper());
-    api.disconnect(client);
-
-    assertThat(api.getLocationCallbacks().get(otherClient)).isNotNull();
-  }
-
   @Test public void requestLocationUpdates_listener_shouldReturnFusedLocationPendingResult() {
     TestLocationListener listener = new TestLocationListener();
     PendingResult<Status> result = api.requestLocationUpdates(client, LocationRequest.create(),

--- a/lost/src/test/java/com/mapzen/android/lost/internal/FusedLocationServiceConnectionManagerTest.java
+++ b/lost/src/test/java/com/mapzen/android/lost/internal/FusedLocationServiceConnectionManagerTest.java
@@ -1,7 +1,5 @@
 package com.mapzen.android.lost.internal;
 
-import com.mapzen.android.lost.api.LostApiClient;
-
 import org.junit.Test;
 
 import android.content.Context;
@@ -82,7 +80,7 @@ public class FusedLocationServiceConnectionManagerTest {
 
   @Test public void disconnect_shouldNotSetStateConnectingConnected() {
     connectionManager.connect(null, null);
-    connectionManager.disconnect(null);
+    connectionManager.disconnect();
     assertThat(connectionManager.isConnecting()).isFalse();
     assertThat(connectionManager.isConnected()).isFalse();
   }
@@ -91,7 +89,7 @@ public class FusedLocationServiceConnectionManagerTest {
     TestEventCallbacks callbacks = new TestEventCallbacks();
     connectionManager.setEventCallbacks(callbacks);
     connectionManager.connect(null, null);
-    connectionManager.disconnect(null);
+    connectionManager.disconnect();
     assertThat(callbacks.connected).isFalse();
   }
 
@@ -99,7 +97,7 @@ public class FusedLocationServiceConnectionManagerTest {
     TestEventCallbacks callbacks = new TestEventCallbacks();
     connectionManager.setEventCallbacks(callbacks);
     connectionManager.connect(null, null);
-    connectionManager.disconnect(null);
+    connectionManager.disconnect();
     assertThat(callbacks.idleOnDisconnect).isTrue();
   }
 
@@ -107,7 +105,7 @@ public class FusedLocationServiceConnectionManagerTest {
     TestEventCallbacks callbacks = new TestEventCallbacks();
     callbacks.onConnect(null);
     connectionManager.setEventCallbacks(callbacks);
-    connectionManager.disconnect(null);
+    connectionManager.disconnect();
     assertThat(callbacks.connected).isTrue();
   }
 
@@ -206,7 +204,7 @@ public class FusedLocationServiceConnectionManagerTest {
     }
 
     @Override
-    public void onDisconnect(LostApiClient client, boolean disconnectService) {
+    public void onDisconnect() {
       connected = false;
       idleOnDisconnect = !connectionManager.isConnected() && !connectionManager.isConnecting();
     }

--- a/lost/src/test/java/com/mapzen/android/lost/internal/LostClientManagerTest.java
+++ b/lost/src/test/java/com/mapzen/android/lost/internal/LostClientManagerTest.java
@@ -198,28 +198,28 @@ public class LostClientManagerTest {
     assertThat(manager.hasNoListeners()).isFalse();
   }
 
-  @Test public void disconnect_shouldRemoveListenersForClient() {
+  @Test public void removeClient_shouldRemoveListenersForClient() {
     LocationRequest request = LocationRequest.create();
     TestLocationListener listener = new TestLocationListener();
     manager.addListener(client, request, listener);
-    manager.disconnect(client);
+    manager.removeClient(client);
     assertThat(manager.getLocationListeners().get(client)).isNull();
   }
 
-  @Test public void disconnect_shouldRemovePendingIntentsForClient() {
+  @Test public void removeClient_shouldRemovePendingIntentsForClient() {
     LocationRequest request = LocationRequest.create();
     PendingIntent pendingIntent = mock(PendingIntent.class);
     manager.addPendingIntent(client, request, pendingIntent);
-    manager.disconnect(client);
+    manager.removeClient(client);
     assertThat(manager.getPendingIntents().get(client)).isNull();
   }
 
-  @Test public void disconnect_shouldRemoveCallbacksForClient() {
+  @Test public void removeClient_shouldRemoveCallbacksForClient() {
     LocationRequest request = LocationRequest.create();
     TestLocationCallback callback = new TestLocationCallback();
     Looper looper = mock(Looper.class);
     manager.addLocationCallback(client, request, callback, looper);
-    manager.disconnect(client);
+    manager.removeClient(client);
     assertThat(manager.getLocationCallbacks().get(client)).isNull();
   }
 

